File: extensions/pi-looper/automations/worker-watch-decision.py
#!/usr/bin/env python3
"""Decide whether issue-coordinator should keep watching a Worker.

The promise file remains the only completion authority. This helper only decides
whether a missing/invalid promise is safe to nudge, keep waiting, or consider for
pane close based on recent activity and the post-nudge grace period.
"""

from __future__ import annotations

import argparse
import json
import math
import sys
from datetime import datetime, timezone
from typing import Any, Iterable

RECENT_WORKER_ACTIVITY_SECONDS = 600
MIN_POST_NUDGE_GRACE_SECONDS = 300
ACTIVE_AGENT_STATUSES = frozenset({"active", "busy", "running", "working"})
SETTLED_PROMISE_STATUSES = frozenset({"complete", "blocked"})

TOP_LEVEL_ACTIVITY_KEYS = (
    "lastActivityAt",
    "lastToolActivityAt",
    "lastFileReadAt",
    "lastAgentSessionUpdatedAt",
    "agentUpdatedAt",
    "sessionUpdatedAt",
    "paneUpdatedAt",
    "recentOutputAt",
)
NESTED_ACTIVITY_KEYS = ("lastActivityAt", "updatedAt", "lastUpdatedAt", "lastOutputAt")
NESTED_ACTIVITY_OBJECTS = ("agent", "session", "pane")
SESSION_ACTIVITY_KINDS = frozenset({"agent_session", "session"})
PANE_OUTPUT_ACTIVITY_KINDS = frozenset({"pane_output", "output"})


def parse_time(value: Any) -> datetime | None:
    if not value:
        return None
    if isinstance(value, (int, float)):
        return datetime.fromtimestamp(value, tz=timezone.utc)
    try:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        return parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def seconds_between(now: datetime, past: datetime | None) -> int | None:
    if past is None:
        return None
    return math.floor((now - past).total_seconds())


def iter_activity_times(observation: dict[str, Any]) -> Iterable[datetime]:
    for key in TOP_LEVEL_ACTIVITY_KEYS:
        parsed = parse_time(observation.get(key))
        if parsed is not None:
            yield parsed

    for object_key in NESTED_ACTIVITY_OBJECTS:
        nested = observation.get(object_key)
        if not isinstance(nested, dict):
            continue
        for key in NESTED_ACTIVITY_KEYS:
            parsed = parse_time(nested.get(key))
            if parsed is not None:
                yield parsed

    for event in observation.get("activity") or observation.get("events") or []:
        if not isinstance(event, dict):
            continue
        parsed = parse_time(event.get("at") or event.get("createdAt") or event.get("updatedAt"))
        if parsed is not None:
            yield parsed


def latest_activity_at(observation: dict[str, Any]) -> datetime | None:
    times = list(iter_activity_times(observation))
    return max(times) if times else None


def active_agent_status(observation: dict[str, Any]) -> str:
    direct = observation.get("agentStatus") or observation.get("agent_status") or observation.get("status")
    if direct:
        return str(direct).lower()
    agent = observation.get("agent")
    if isinstance(agent, dict):
        return str(agent.get("agent_status") or agent.get("status") or "").lower()
    return ""


def activity_kinds(observation: dict[str, Any]) -> set[str]:
    kinds: set[str] = set()
    for event in observation.get("activity") or observation.get("events") or []:
        if isinstance(event, dict) and event.get("kind"):
            kinds.add(str(event["kind"]))
    return kinds


def has_any_key(data: dict[str, Any], keys: Iterable[str]) -> bool:
    return any(key in data for key in keys)


def has_checked_nested_object(observation: dict[str, Any], object_key: str) -> bool:
    nested = observation.get(object_key)
    return isinstance(nested, dict) and has_any_key(nested, NESTED_ACTIVITY_KEYS)


def missing_required_observations(observation: dict[str, Any]) -> list[str]:
    missing: list[str] = []
    kinds = activity_kinds(observation)
    if not active_agent_status(observation):
        missing.append("agent_status")
    if not (
        has_any_key(observation, ("lastAgentSessionUpdatedAt", "agentUpdatedAt", "sessionUpdatedAt"))
        or has_checked_nested_object(observation, "agent")
        or has_checked_nested_object(observation, "session")
        or bool(kinds & SESSION_ACTIVITY_KINDS)
    ):
        missing.append("agent_session_updated_at")
    if not (
        has_any_key(observation, ("paneUpdatedAt", "recentOutputAt"))
        or has_checked_nested_object(observation, "pane")
        or bool(kinds & PANE_OUTPUT_ACTIVITY_KINDS)
    ):
        missing.append("pane_recent_output")
    return missing


def parse_bool(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value.lower() in {"1", "true", "yes", "on"}
    return bool(value)


def build_observations(
    observation: dict[str, Any],
    now: datetime,
    last_activity: datetime | None,
    nudge_sent_at: datetime | None,
) -> dict[str, Any]:
    return {
        "promiseStatus": str(observation.get("promiseStatus") or observation.get("promise_status") or "none"),
        "worktreeHasChanges": parse_bool(observation.get("worktreeHasChanges", False)),
        "agentStatus": active_agent_status(observation),
        "lastActivityAt": last_activity.isoformat().replace("+00:00", "Z") if last_activity else None,
        "secondsSinceLastActivity": seconds_between(now, last_activity),
        "nudgeSentAt": nudge_sent_at.isoformat().replace("+00:00", "Z") if nudge_sent_at else None,
        "secondsSinceNudge": seconds_between(now, nudge_sent_at),
        "recentWorkerActivitySeconds": RECENT_WORKER_ACTIVITY_SECONDS,
        "minPostNudgeGraceSeconds": MIN_POST_NUDGE_GRACE_SECONDS,
    }


def close_log(observations: dict[str, Any]) -> str:
    return (
        f"promise={observations['promiseStatus']}; "
        f"agentStatus={observations['agentStatus'] or 'unknown'}; "
        f"lastActivityAt={observations['lastActivityAt'] or 'unknown'}; "
        f"secondsSinceLastActivity={observations['secondsSinceLastActivity']}; "
        f"nudgeSentAt={observations['nudgeSentAt'] or 'unknown'}; "
        f"secondsSinceNudge={observations['secondsSinceNudge']}; "
        f"minPostNudgeGraceSeconds={observations['minPostNudgeGraceSeconds']}; "
        f"worktreeHasChanges={str(observations['worktreeHasChanges']).lower()}"
    )


def decide_worker_watch(observation: dict[str, Any]) -> dict[str, Any]:
    now = parse_time(observation.get("now")) or datetime.now(timezone.utc)
    promise_status = str(observation.get("promiseStatus") or observation.get("promise_status") or "none")
    last_activity = latest_activity_at(observation)
    nudge_sent_at = parse_time(observation.get("nudgeSentAt") or observation.get("lastNudgeAt"))
    observations = build_observations(observation, now, last_activity, nudge_sent_at)

    if promise_status in SETTLED_PROMISE_STATUSES:
        return {"action": "promise_settled", "reason": promise_status, "observations": observations}

    agent_status = observations["agentStatus"]
    if agent_status in ACTIVE_AGENT_STATUSES:
        return {"action": "continue_waiting", "reason": "agent_status_active", "observations": observations}

    activity_age = observations["secondsSinceLastActivity"]
    if activity_age is not None and activity_age <= RECENT_WORKER_ACTIVITY_SECONDS:
        return {"action": "continue_waiting", "reason": "recent_activity", "observations": observations}

    if nudge_sent_at is None:
        return {"action": "nudge_worker", "reason": "missing_promise", "observations": observations}

    nudge_age = observations["secondsSinceNudge"]
    if nudge_age is None or nudge_age < MIN_POST_NUDGE_GRACE_SECONDS:
        return {"action": "continue_waiting", "reason": "nudge_grace_period", "observations": observations}

    missing_observations = missing_required_observations(observation)
    if missing_observations:
        return {
            "action": "collect_observations",
            "reason": "missing_required_observations",
            "missingObservations": missing_observations,
            "observations": observations,
        }

    return {
        "action": "may_close_pane",
        "reason": "inactive_after_grace",
        "observations": observations,
        "closeLog": close_log(observations),
    }


def load_input(path: str | None) -> dict[str, Any]:
    with (open(path, "r", encoding="utf-8") if path else sys.stdin) as stream:
        data = json.load(stream)
    if not isinstance(data, dict):
        raise ValueError("watch observation JSON must be an object")
    return data


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", help="Path to watch observation JSON. Defaults to stdin.")
    parser.add_argument("--now", help="Override observation now for deterministic runs.")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    observation = load_input(args.input)
    if args.now:
        observation["now"] = args.now
    print(json.dumps(decide_worker_watch(observation), ensure_ascii=False, sort_keys=True))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(f"worker-watch-decision.py: {error}", file=sys.stderr)
        raise SystemExit(2)

File: extensions/deadloop/automations/worker-watch-decision.ts
#!/usr/bin/env node
// Decide whether issue-coordinator should keep watching a Worker. CommonJS-shaped
// so it can run directly with `node worker-watch-decision.ts`.

const fs = require("node:fs") as typeof import("node:fs");

type WorkerWatchObservation = Record<string, any>;

const RECENT_WORKER_ACTIVITY_SECONDS = 600;
const MIN_POST_NUDGE_GRACE_SECONDS = 300;
const ACTIVE_AGENT_STATUSES = new Set(["active", "busy", "running", "working"]);
const SETTLED_PROMISE_STATUSES = new Set(["complete", "blocked"]);

const TOP_LEVEL_ACTIVITY_KEYS = [
  "lastActivityAt",
  "lastToolActivityAt",
  "lastFileReadAt",
  "lastAgentSessionUpdatedAt",
  "agentUpdatedAt",
  "sessionUpdatedAt",
  "paneUpdatedAt",
  "recentOutputAt",
];
const NESTED_ACTIVITY_KEYS = ["lastActivityAt", "updatedAt", "lastUpdatedAt", "lastOutputAt"];
const NESTED_ACTIVITY_OBJECTS = ["agent", "session", "pane"];
const SESSION_ACTIVITY_KINDS = new Set(["agent_session", "session"]);
const PANE_OUTPUT_ACTIVITY_KINDS = new Set(["pane_output", "output"]);

function parseTimeForWorkerWatch(value: unknown): Date | null {
  if (!value) return null;
  if (typeof value === "number") {
    const parsed = new Date(value * 1000);
    return Number.isNaN(parsed.getTime()) ? null : parsed;
  }
  const text = String(value);
  const normalized = /^\d{4}-\d{2}-\d{2}T/.test(text) && !/(?:Z|[+-]\d{2}:?\d{2})$/.test(text) ? `${text}Z` : text;
  const parsed = new Date(normalized);
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

function secondsBetweenForWorkerWatch(now: Date, past: Date | null): number | null {
  if (!past) return null;
  return Math.floor((now.getTime() - past.getTime()) / 1000);
}

function iterActivityTimesForWorkerWatch(observation: WorkerWatchObservation): Date[] {
  const times: Date[] = [];
  for (const key of TOP_LEVEL_ACTIVITY_KEYS) {
    const parsed = parseTimeForWorkerWatch(observation[key]);
    if (parsed) times.push(parsed);
  }

  for (const objectKey of NESTED_ACTIVITY_OBJECTS) {
    const nested = observation[objectKey];
    if (!nested || typeof nested !== "object" || Array.isArray(nested)) continue;
    for (const key of NESTED_ACTIVITY_KEYS) {
      const parsed = parseTimeForWorkerWatch(nested[key]);
      if (parsed) times.push(parsed);
    }
  }

  for (const event of observation.activity || observation.events || []) {
    if (!event || typeof event !== "object") continue;
    const parsed = parseTimeForWorkerWatch(event.at || event.createdAt || event.updatedAt);
    if (parsed) times.push(parsed);
  }
  return times;
}

function latestActivityAtForWorkerWatch(observation: WorkerWatchObservation): Date | null {
  const times = iterActivityTimesForWorkerWatch(observation);
  if (!times.length) return null;
  return new Date(Math.max(...times.map((time) => time.getTime())));
}

function activeAgentStatusForWorkerWatch(observation: WorkerWatchObservation): string {
  const direct = observation.agentStatus || observation.agent_status || observation.status;
  if (direct) return String(direct).toLowerCase();
  const agent = observation.agent;
  if (agent && typeof agent === "object") return String(agent.agent_status || agent.status || "").toLowerCase();
  return "";
}

function activityKindsForWorkerWatch(observation: WorkerWatchObservation): Set<string> {
  const kinds = new Set<string>();
  for (const event of observation.activity || observation.events || []) {
    if (event && typeof event === "object" && event.kind) kinds.add(String(event.kind));
  }
  return kinds;
}

function hasAnyWorkerWatchKey(data: WorkerWatchObservation, keys: string[]): boolean {
  return keys.some((key) => Object.prototype.hasOwnProperty.call(data, key));
}

function hasCheckedWorkerWatchNestedObject(observation: WorkerWatchObservation, objectKey: string): boolean {
  const nested = observation[objectKey];
  return Boolean(nested && typeof nested === "object" && hasAnyWorkerWatchKey(nested, NESTED_ACTIVITY_KEYS));
}

function setIntersectsForWorkerWatch(left: Set<string>, right: Set<string>): boolean {
  for (const value of left) if (right.has(value)) return true;
  return false;
}

function missingRequiredObservationsForWorkerWatch(observation: WorkerWatchObservation): string[] {
  const missing: string[] = [];
  const kinds = activityKindsForWorkerWatch(observation);
  if (!activeAgentStatusForWorkerWatch(observation)) missing.push("agent_status");
  if (
    !(
      hasAnyWorkerWatchKey(observation, ["lastAgentSessionUpdatedAt", "agentUpdatedAt", "sessionUpdatedAt"]) ||
      hasCheckedWorkerWatchNestedObject(observation, "agent") ||
      hasCheckedWorkerWatchNestedObject(observation, "session") ||
      setIntersectsForWorkerWatch(kinds, SESSION_ACTIVITY_KINDS)
    )
  ) {
    missing.push("agent_session_updated_at");
  }
  if (
    !(
      hasAnyWorkerWatchKey(observation, ["paneUpdatedAt", "recentOutputAt"]) ||
      hasCheckedWorkerWatchNestedObject(observation, "pane") ||
      setIntersectsForWorkerWatch(kinds, PANE_OUTPUT_ACTIVITY_KINDS)
    )
  ) {
    missing.push("pane_recent_output");
  }
  return missing;
}

function parseBoolForWorkerWatch(value: unknown): boolean {
  if (typeof value === "boolean") return value;
  if (typeof value === "string") return ["1", "true", "yes", "on"].includes(value.toLowerCase());
  return Boolean(value);
}

function isoForWorkerWatch(value: Date | null): string | null {
  return value ? value.toISOString().replace(".000Z", "Z") : null;
}

function buildWorkerWatchObservations(
  observation: WorkerWatchObservation,
  now: Date,
  lastActivity: Date | null,
  nudgeSentAt: Date | null,
): WorkerWatchObservation {
  return {
    promiseStatus: String(observation.promiseStatus || observation.promise_status || "none"),
    worktreeHasChanges: parseBoolForWorkerWatch(observation.worktreeHasChanges ?? false),
    agentStatus: activeAgentStatusForWorkerWatch(observation),
    lastActivityAt: isoForWorkerWatch(lastActivity),
    secondsSinceLastActivity: secondsBetweenForWorkerWatch(now, lastActivity),
    nudgeSentAt: isoForWorkerWatch(nudgeSentAt),
    secondsSinceNudge: secondsBetweenForWorkerWatch(now, nudgeSentAt),
    recentWorkerActivitySeconds: RECENT_WORKER_ACTIVITY_SECONDS,
    minPostNudgeGraceSeconds: MIN_POST_NUDGE_GRACE_SECONDS,
  };
}

function closeLogForWorkerWatch(observations: WorkerWatchObservation): string {
  return [
    `promise=${observations.promiseStatus}`,
    `agentStatus=${observations.agentStatus || "unknown"}`,
    `lastActivityAt=${observations.lastActivityAt || "unknown"}`,
    `secondsSinceLastActivity=${observations.secondsSinceLastActivity}`,
    `nudgeSentAt=${observations.nudgeSentAt || "unknown"}`,
    `secondsSinceNudge=${observations.secondsSinceNudge}`,
    `minPostNudgeGraceSeconds=${observations.minPostNudgeGraceSeconds}`,
    `worktreeHasChanges=${String(observations.worktreeHasChanges).toLowerCase()}`,
  ].join("; ");
}

function decideWorkerWatch(observation: WorkerWatchObservation): WorkerWatchObservation {
  const now = parseTimeForWorkerWatch(observation.now) || new Date();
  const promiseStatus = String(observation.promiseStatus || observation.promise_status || "none");
  const lastActivity = latestActivityAtForWorkerWatch(observation);
  const nudgeSentAt = parseTimeForWorkerWatch(observation.nudgeSentAt || observation.lastNudgeAt);
  const observations = buildWorkerWatchObservations(observation, now, lastActivity, nudgeSentAt);

  if (SETTLED_PROMISE_STATUSES.has(promiseStatus)) {
    return { action: "promise_settled", reason: promiseStatus, observations };
  }

  const agentStatus = observations.agentStatus;
  if (ACTIVE_AGENT_STATUSES.has(agentStatus)) {
    return { action: "continue_waiting", reason: "agent_status_active", observations };
  }

  const activityAge = observations.secondsSinceLastActivity;
  if (activityAge !== null && activityAge <= RECENT_WORKER_ACTIVITY_SECONDS) {
    return { action: "continue_waiting", reason: "recent_activity", observations };
  }

  if (!nudgeSentAt) {
    return { action: "nudge_worker", reason: "missing_promise", observations };
  }

  const nudgeAge = observations.secondsSinceNudge;
  if (nudgeAge === null || nudgeAge < MIN_POST_NUDGE_GRACE_SECONDS) {
    return { action: "continue_waiting", reason: "nudge_grace_period", observations };
  }

  const missingObservations = missingRequiredObservationsForWorkerWatch(observation);
  if (missingObservations.length) {
    return {
      action: "collect_observations",
      reason: "missing_required_observations",
      missingObservations,
      observations,
    };
  }

  return {
    action: "may_close_pane",
    reason: "inactive_after_grace",
    observations,
    closeLog: closeLogForWorkerWatch(observations),
  };
}

function loadWorkerWatchInput(file: string | undefined): WorkerWatchObservation {
  const data = JSON.parse(file ? fs.readFileSync(file, "utf8") : fs.readFileSync(0, "utf8"));
  if (!data || typeof data !== "object" || Array.isArray(data)) throw new Error("watch observation JSON must be an object");
  return data;
}

function requiredWorkerWatchValue(argv: string[], index: number, flag: string): string {
  const value = argv[index + 1];
  if (!value || value.startsWith("--")) throw new Error(`${flag} requires a value`);
  return value;
}

function parseWorkerWatchArgs(argv: string[]): WorkerWatchObservation {
  const parsed: WorkerWatchObservation = {};
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === "--input") {
      parsed.input = requiredWorkerWatchValue(argv, index, token);
      index += 1;
      continue;
    }
    if (token === "--now") {
      parsed.now = requiredWorkerWatchValue(argv, index, token);
      index += 1;
      continue;
    }
    if (token === "--help" || token === "-h") {
      parsed.help = true;
      continue;
    }
    throw new Error(`unknown flag: ${token}`);
  }
  return parsed;
}

function workerWatchHelp(): string {
  return "Usage: worker-watch-decision.ts [--input FILE] [--now ISO_TIMESTAMP]";
}

function main(argv: string[] = process.argv.slice(2)): number {
  const args = parseWorkerWatchArgs(argv);
  if (args.help) {
    process.stdout.write(`${workerWatchHelp()}\n`);
    return 0;
  }
  const observation = loadWorkerWatchInput(args.input);
  if (args.now) observation.now = args.now;
  process.stdout.write(`${JSON.stringify(decideWorkerWatch(observation))}\n`);
  return 0;
}

if (require.main === module) {
  try {
    process.exitCode = main();
  } catch (error) {
    console.error(`worker-watch-decision.ts: ${error instanceof Error ? error.message : String(error)}`);
    process.exitCode = 2;
  }
}

module.exports = { decideWorkerWatch };
